### PR TITLE
Key configuration

### PR DESCRIPTION
**Pull request by peppy**
Currently only available as a VisualTest. PRing for review and feedback.

- [x] Depends on https://github.com/ppy/osu-framework/pull/965

### CODE REVIEW

**Comment by peppy** [comment]
Had to combine https://github.com/ppy/osu/pull/1107 into this PR due to multiple framework changes in wrong order

File: osu.Game/Input/Bindings/DatabasedKeyBinding.cs [hunk 1]
```diff
@@ -24,11 +27,12 @@ public string KeysString
             private set { KeyCombination = value; }
         }
 
+        [Indexed]
         [Column("Action")]
```
Comment (Tom94) [inline osu.Game/Input/Bindings/DatabasedKeyBinding.cs]: I would prefer if you didn't use `new` and instead used a different name such as `DatabasedAction`.

File: osu.Game/Input/Bindings/DatabasedKeyBindingInputManager.cs [hunk 2]
```diff
@@ -18,7 +19,9 @@ public abstract class DatabasedKeyBindingInputManager<T> : KeyBindingInputManage
 
         private readonly int? variant;
 
-        private BindingStore store;
+        private KeyBindingStore store;
+
+        public override IEnumerable<KeyBinding> DefaultMappings => ruleset.CreateInstance().GetDefaultKeyBindings();
```
Comment (Tom94) [inline osu.Game/Input/Bindings/DatabasedKeyBindingInputManager.cs]: In one place it's called "keybindings" and in one "mappings". That seems strange.

File: osu.Game/Input/KeyBindingStore.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                        Reset();
+                        break;
+                }
+            }
+        }
+
+        protected override void Prepare(bool reset = false)
+        {
+            if (reset)
+            {
```
Comment (Tom94) [inline osu.Game/Input/KeyBindingStore.cs]: No curly braces.

File: osu.Game/Input/KeyBindingStore.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        protected override Type[] ValidTypes => new[]
+        {
+            typeof(DatabasedKeyBinding)
+        };
+
+        public IEnumerable<KeyBinding> Query(int? rulesetId = null, int? variant = null) =>
+            Query<DatabasedKeyBinding>(b => b.RulesetID == rulesetId && b.Variant == variant);
+
+        public void Update(KeyBinding keyBinding)
+        {
```
Comment (Tom94) [inline osu.Game/Input/KeyBindingStore.cs]: =>

File: osu.Game/OsuGameBase.cs [hunk 5]
```diff
@@ -183,12 +183,14 @@ protected override void LoadComplete()
...
 
+            GlobalBindingInputManager globalBinding;
+
             base.Content.Add(new RatioAdjust
             {
                 Children = new Drawable[]
                 {
                     Cursor = new MenuCursor(),
-                    new GlobalBindingInputManager(this)
+                    globalBinding = new GlobalBindingInputManager(this)
```
Comment (Tom94) [inline osu.Game/OsuGameBase.cs]: `Binding` -> `KeyBinding`

File: osu.Game/Overlays/KeyBinding/GlobalBindingsSection.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using osu.Framework.Input.Bindings;
+using osu.Game.Graphics;
+
+namespace osu.Game.Overlays.KeyBinding
+{
+    public class GlobalBindingsSection : KeyBindingsSection
```
Comment (Tom94) [inline osu.Game/Overlays/KeyBinding/GlobalBindingsSection.cs]: KeyBindings

File: osu.Game/Overlays/KeyBinding/KeyBindingRow.cs [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+                }
+            }
+
+            public KeyButton(Framework.Input.Bindings.KeyBinding keyBinding)
+            {
+                KeyBinding = keyBinding;
+
+                Margin = new MarginPadding(padding);
+
+                //var isDefault = keyBinding.Action is Enum;
```
Comment (Tom94) [inline osu.Game/Overlays/KeyBinding/KeyBindingRow.cs]: Should be removed.

**Review by Tom94 — DISMISSED** [review]

**Review by Tom94 — APPROVED** [review]

**Pull request merged**